File: practice/algorithms/implementation/strange_counter/StrangeCounter.js
'use strict';

const processData = input => {
    const t = parseInt(input);
    const cycleEndTime = cycleEnd(t, 0, 3);
    const timeDiff = cycleEndTime - t;
    const finalValue = timeDiff + 1;
    console.log(finalValue);
};

const cycleEnd = (time, start, step) => {
    if(start > time) {
        return start;
    }
    return cycleEnd(time, start + step, step * 2);
};

process.stdin.resume();
process.stdin.setEncoding("ascii");

let _input = "";
process.stdin.on("data", input => _input += input);
process.stdin.on("end", () => processData(_input));

File: practice/algorithms/implementation/strange_counter/strange_counter.rb
def cycle_end(time, start, step)
  return start if(start > time)
  cycle_end(time, start + step, step * 2)
end

t = gets.to_i
cycle_end_time = cycle_end(t, 0, 3)
time_diff = cycle_end_time - t
final_value = time_diff + 1
puts final_value